modules/auxiliary/gather/ldap_query.rb

```diff
@@ -206,15 +301,113 @@ def output_data_csv(entries)
     generate_rex_tables(entries, 'csv')
   end
 
-  def normalize_entries(entries)
+  def query_attributes_data(ldap, entry_keys)
+    # def perform_ldap_query(ldap, filter, attributes, base: nil)
+    filter = '(|'
+    entry_keys.each do |key|
+      filter += "(LDAPDisplayName=#{key})"
+    end
+    filter += ')'
+    attributes = ['LDAPDisplayName', 'isSingleValued', 'oMSyntax', 'attributeSyntax']
+    attributes_data = perform_ldap_query(ldap, filter, attributes, base: ['CN=Schema,CN=Configuration', @base_dn].join(','))
+
+    entry_list = []
+    for entry in attributes_data do
+      data = {}
+      data[:ldapdisplayname] = entry[:ldapdisplayname][0]
+      if entry[:issinglevalued][0] == 'TRUE'
+        data[:issinglevalued] = 1
+      else
+        data[:issinglevalued] = 0
+      end
```

Comment:
Why not just use a boolean here?

```suggestion
      data[:issinglevalued] = entry[:issinglevalued][0] == 'TRUE'
```

Also I don't see where this is used anywhere. Shouldn't it be used to convert the returned arrays to a single value?

Reply:
In theory it just means that "When this flag is present on an [AttributeType](https://ldapwiki.com/wiki/AttributeTypes) definition the attribute is restricted to possessing a single value. Absence of the flag implies the [AttributeType](https://ldapwiki.com/wiki/AttributeTypes) definition may have multiple values." according to https://ldapwiki.com/wiki/SINGLE-VALUE.

Though the part that is a little confusing is if it means multiple types of data or just multiple values of the same type. If later than your point is right. Just need to try find some experiments and play around with this a bit more.

Reply:
So ironically enough we already have this for `cacertificate` and `certificatetemplates`. I'm not sure we are appropriately handling `cacertificate` but for the case of `certificatetemplates`, since it is an array and we are using `.map` we incidentally end up handling all of its data in the various array elements successfully.

We will likely still need to see if any of the other data types we support are labeled as `isSingleValued = False` and update accordingly.

Reply:
Using booleans fixed with https://github.com/rapid7/metasploit-framework/pull/16979/commits/be95e69e722cf8950502f5c5562ba2e76d01c872 however the underlying issue of not using `isSingleValued` correctly remains and will be fixed in a future commit.

Reply:
Should now be fixed in https://github.com/rapid7/metasploit-framework/pull/16979/commits/6880928d88182b51f532b884dd86622a27a69e25 though may still need to potentially remove the `isSingleValued` variable from the code if we aren't going to be using it anymore.